packages/sandcastle/src/Gallery.tsx

```diff
@@ -15,6 +16,140 @@ export type GalleryItem = {
+        return;
+      }
+      const searchString = searchTerm;
+      if (!searchString || searchString === "") {
+        setSearchResults(null);
+        return;
+      }
+      const thisRequestId = ++requestId.current;
+
+      const search = await pagefind.current.search(searchString, {
+        filters: {
+          tags: currentTag === "All" ? undefined : currentTag,
+        },
+      });
+
+      const resultDataPromises = await Promise.allSettled(
+        search.results.map((result) => {
+          return result.data();
+        }),
+      );
+      const resultData = resultDataPromises
+        .map((result) => {
+          return result.status === "fulfilled" ? result.value : undefined;
+        })
+        .filter((result) => !!result);
+
+      if (thisRequestId === requestId.current) {
+        setSearchResults(resultData);
+      }
+    }
+
+    doSearch();
+  }, [searchTerm, currentTag, setSearchResults]);
+
+  useEffect(() => {
+    async function loadPageFind() {
+      const pagefindImport: Pagefind = await import(
+        // vite doesn't like the dynamic import
+        /* @vite-ignore */
+        `${GALLERY_BASE}/pagefind/pagefind.js`
+      );
+
+      pagefindImport.init();
+      await pagefindImport.options({
+        baseUrl: getBaseUrl(),
+      });
+      pagefind.current = pagefindImport;
+
+      const filters = await pagefind.current.filters();
+      setAvailableFilters(filters);
+
+      setPagefindLoaded(true);
+    }
+    if (isFirstLoad.current) {
+      // This prevents double activation when using StrictMode
+      isFirstLoad.current = false;
+      loadPageFind();
+    }
+  }, []);
+
+  const debounceTimeout = useRef<NodeJS.Timeout>(null);
+
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  return (
+    <>
+      {!!searchTerm && (
+        <Button
+          onClick={() => {
+            setSearchTerm("");
+            if (inputRef.current) {
+              inputRef.current.value = "";
+            }
+          }}
+        >
+          Clear
+        </Button>
+      )}
+      <Input
+        ref={inputRef}
+        disabled={!pagefindLoaded}
+        onChange={(e) => {
+          if (debounceTimeout.current) {
+            clearInterval(debounceTimeout.current);
+          }
+          const newSearchTerm = e.target.value;
+          pagefind.current?.preload(newSearchTerm);
+          debounceTimeout.current = setTimeout(() => {
+            setSearchTerm(newSearchTerm);
+          }, 300);
+        }}
+        placeholder="Filter demos..."
+      />
+      <Select.Root>
+        <Select.HtmlSelect
+          disabled={!pagefindLoaded}
+          onChange={(e) => {
+            setCurrentTag(e.target.value);
+            setTag(e.target.value);
+          }}
+        >
+          <option value={"All"}>All</option>
+          {Object.keys(availableFilters.tags).map((label) => (
+            <option value={label} key={label}>
+              {label}
+            </option>
+          ))}
```

Comment:
I think it may end up being a bit much to iterate on for this particular PR, but I think we discussed some alternatives for how the categories/labels are presented offline. 

Could you please add "Refine gallery filter by label presentation" to the list of TODO's?